reconciler/leader.go

```diff
@@ -0,0 +1,118 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reconciler
+
+import (
+	"sync"
+
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// Bucket is an opaque type used to scope leadership.
+type Bucket interface {
+	// Name returns a string representing this bucket, which uniquely
+	// identifies the bucket and is suitable for use as a resource lock name.
+	Name() string
+
+	// Has determines whether this Bucket contains a particular key.
+	Has(key types.NamespacedName) bool
+}
+
+// UniversalBucket returns a Bucket that "Has()" all keys.
+func UniversalBucket() Bucket {
+	return &bucket{}
+}
+
+type bucket struct{}
+
+var _ Bucket = (*bucket)(nil)
+
+// Name implements Bucket
+func (b *bucket) Name() string {
+	return ""
+}
+
+// Has implements Bucket
+func (b *bucket) Has(nn types.NamespacedName) bool {
+	return true
+}
+
+// LeaderAware is implemented by Reconcilers that are aware of their leader status.
+type LeaderAware interface {
+	// Promote is called when we become the leader of a given Bucket.  It must be
+	// supplied with an enqueue function through which a Bucket resync may be triggered.
+	Promote(b Bucket, enq func(Bucket, types.NamespacedName))
```

Comment:
I'm still on the fence about the `enq` callback 

`NewReconciler` generated code seems to use it because they're no higher awareness of a workqueue. That could potentially be solved by passing in a `EnqueueKey` function. `controller.NewImpl` is a similar story.

All this feels like a work around because our controller.Impl construction is coupled with the workqueue. It's a chicken and egg problem.  Reconcilers want an enqueue func, but controllers need a reconciler but only when creating a controller do we create a work queue.

If we moved the responsibility of creating the queue outside of the controller I wonder if this gives us the flexibility we want.  Then `MaybeEnqueueBucketKey(reconciler.Bucket, types.NamespacedName)` would only need to a helper and not passed around everywhere.

Reply:
I'm open to signature changes, but I think the more general conversation of the workqueue's relationship to controller.Impl is probably an orthogonal conversation (I'll reserve further comment for that convo 😉 ).

I think my preference would be to land this and iterate rather than trying to explode everything here/now. 😬

Reply:
> I think my preference would be to land this and iterate rather than trying to explode everything here/now. 😬

Agreed - given this is an alpha API we have flexibilty